File: Go/5_longestPalindrome/5_longestPalindrome.go
package q005

import (
	"fmt"
)

func longestPalindrome(src string) string {
	dst := ""

	for i := 0; i < len(src); i++ {
		l1 := expandAroundCenter(src, i, i)
		l2 := expandAroundCenter(src, i, i + 1)

		l := l1
		if l2 > l {
			l = l2
		}

		if l > 1 && l > len(dst) {
			start := i - (l - 1) / 2
			dst = src[start : start + l]
		}
	}

	return dst
}

func expandAroundCenter(src string, left, right int) int {
	for left >= 0 && right < len(src) {
		if src[left] != src[right] {
			break
		}

		left--
		right++
	}

	return right - left - 1
}

// Test ...
func Test() {
	fmt.Println("5_longestPalindrome begin")
	str := "abcccabbaabcba"
	fmt.Println(longestPalindrome(str))
	fmt.Println("5_longestPalindrome end")
}

File: CPP/5_longestPalindrome/5_longestPalindrome.cpp
// 5_longestPalindrome.cpp : 定义控制台应用程序的入口点。
//

#include "stdafx.h"
#include <iostream>
#include <string>
using namespace std;

string longestPalindrome(string src) 
{
	string dst = "";

	int len = src.length();
	for (int i = 0; i < len; i++)
	{
		int left = i - 1;
		int right = i + 1;
		while (right < len)
		{
			if (src[right] != src[i])
			{
				break;
			}

			right++;
			i = right - 1;
		}

		while (left >= 0 && right < len)
		{
			if (src[left] != src[right])
			{
				break;
			}

			left--;
			right++;
		}

		int temp = right - left -1;
		if (temp > 1 && temp > dst.length())
		{
			dst = src.substr(left + 1, temp);
		}

		if (right >= len)
		{
			break;
		}
	}

	return dst;
}

int _tmain(int argc, _TCHAR* argv[])
{
	string str = "abcccabbaabcba";
	cout << longestPalindrome(str) << endl;

	system("pause");
	return 0;
}

